### ISSUE TITLE
[feature request] macros.bindSym can accept computed string

### ISSUE THREAD

**Issue by jangko** [nim-lang/Nim#7827]
Current implementation of bindSym can only accept literal string as it's parameter. Workaround for this situation is to create intermediate macro contains bindSym using computed string from first tier macro.

see #6243 #3559 for two tier macros example.

This two tier macros work reliably but the code is not easy to read and modify. It would be a great benefit for macro users if the bindSym can accept computed string besides literal string.

I know this is quite difficult to achieve. Setting up new VM context to evaluate bindSym parameter inside a macro or static block while the VM already running is not an easy task.

**Comment by cooldome** [comment]
As first simplified version it can be easier to implement only a check if symbol with particular name and type exists and not actually making a sym binding.

**Comment by jangko** [comment]
**Initial investigation:**

bindSym is not a true VM operation.

the node before bindSym operation is something like this:
```text
nkCall
   nkSym "bindSym"
   nkStringLit "typeName"
   nkIntLit bindingRule
```
during semcheck pass, semmagic.magicsAfterOverloadResolution will call semBindSym and replace the last two sons of nkCall children with a symbol(opensymchoice, closedsymchoice, sym)

```text
nkCall
   nkSym "bindSym"
   nkSymXXX "typeName"
```

aha, before it entered the VM, the compiler already have the result of the bindSym!

at vmgen, it copied the node to VM constants table, and during VM execution, bindSym opcode copied the node from constants table back to registry table(more or less like a  MOV).

that's why bindSym cannot accept computed string because symbol binding occured before the VM run.

**Challenges:**
There is a clean separation between semcheck pass and VM pass,  at least they have their own execution context. I am still not sure how exactly to pass data between this two context cleanly.

to successfully bind the symbol, VM need to acquire semcheck context and have access to symbol binding procs.

semcheck context can be passed to VM context during VM context creation. But accessing symbol binding procs cannot just importing the modules, because it will become a conflict of circular import. therefore need to move the code around.

**how far we are**
During my experiment, I pass both sempass context and function pointer of symbol binding proc to VM context using pointer casting. Basically I smuggled the reference using "bindSym" symbol field. At VM context the pointer is cast back to sempass context and function pointer.

Of course this is an awful hacky way, but it works. So we are not too far from implementing a working bindSym that can accept computed string.

perhaps the real issue here is about compiler design/source code organization/why not implement it from earlier days???

**Comment by Araq** [comment]
> There is a clean separation between semcheck pass and VM pass, at least they have their own execution context. I am still not sure how exactly to pass data between this two context cleanly.

There is no way and ``bindSym`` should not become a VM operation. Macros cannot access the symbol table by design. You need really good examples in order to convince me this is a good idea.

**Comment by jangko** [comment]
>There is **no way** and bindSym **should not** become a VM operation

> Macros **cannot** access the symbol table **by design**

Frankly, this sound like a law and a regulation, I'm just not sure who wrote it :smile:.
I won't argue, because I don't want to break any law, and I also don't have a very good examples to convince you.

most of the examples I have involved not a string computation, but binding an ident node.
perhaps until we have a really good examples, can you allow me to modify the bindsym to accepts an ident node beside literal string? that would not make bindsym become a VM operation.

**Comment by zah** [comment]
Also, macros already work with the symbol table due to `getAst`.

If we accept the rule that macro evaluation happens in the scope where the macro call is placed, there are few more features that can be given proper semantics. For example, it would be possible to introduce a magic proc `determineType` that could be given an arbitrary AST node. This has been requested in the past and @mratsim have even invented some crazy hacks based on `getAst` that achieve it to some extent. We are in C++ territory here - people "discover" and use hacks instead of deliberately designed language features.

**Comment by Araq** [comment]
> Also, macros already work with the symbol table due to getAst, although this loophole can be closed.

How so? The VM doesn't access the symbol table, it can't, can it?

**Comment by Araq** [comment]
> perhaps until we have really good examples, can you allow me to modify the bindsym to accepts an ident node beside literal string? that would not make bindsym become a VM operation.

Intriguing. How would this solve your problem?

**Comment by jangko** [comment]
```Nim
import macros

type
  Apple = ref object
    name: string
    color: int
    weight: int    
  
proc initApple(name: string): Apple =
  discard
  
proc deinitApple(x: Apple) =
  discard
  
macro wrapObject(obj: typed, n: varargs[untyped]): untyped =
  echo n.treeRepr
  
wrapObject(Apple):
  initApple
  ~deinitApple
  name(get, set)
  color(get)
  weight(set)
```

This is a small example how I made wrapper to embed scripting language using macros.
Because the wrapper using some kind of DSL, I cannot pass it as typed arguments,
as a consequences  I also cannot access the arguments type.

Here the bindSym limitation comes in, I cannot bind 'initApple', 'deinitApple', 'name', 'color', 'weight' directly. I use multi tier macros as a workaround. But implementing multi tier macros every time I want to experiment with some DSL is cumbersome and tedious. The code also become hard to follow.

If the bindSym can bind symbol to an ident node, that would eliminate multi tier macros workaround.
And I believe, given this additional power will trigger user creativity, and they soon will find that bindSym limitation will only hindered they creative process, as they struggle to find yet another workaround.

This is how I hope, we can prepare a way that someday someone comes with a convincing use case, and you can wholeheartedly ready to break the law of this clean phase separation compiler design.

**Comment by jangko** [comment]
> Intriguing. How would this solve your problem?

ah, now I understand, right now it is nearly impossible to implement this feature without any modification to compiler design.

**Comment by Araq** [comment]
Well you can cheat by registering a VM callback that captures ``c: PContext`` in the closure environment.

**Comment by jangko** [comment]
I already try the VM callback.

The closure technique doesn't work. It will make the PContext associated with the callback and there is no way to update the PContext. (or there is????)

each call to bindSym can happen at different module, thus it can have different PContext. We need to pass the PContext along with the AST or some other technique. not sure how. a dummy node with generic payload?
